broker-core/src/main/java/io/zeebe/broker/subscription/message/processor/PublishMessageProcessor.java

```diff
@@ -82,15 +84,23 @@ public void processRecord(
       responseWriter.writeRejectionOnCommand(record, RejectionType.BAD_VALUE, rejectionReason);
 
     } else {
-      final long key = streamWriter.writeNewEvent(MessageIntent.PUBLISHED, record.getValue());
+      final TypedBatchWriter batchWriter = streamWriter.newBatch();
+      final long key = batchWriter.addNewEvent(MessageIntent.PUBLISHED, record.getValue());
       responseWriter.writeEventOnCommand(key, MessageIntent.PUBLISHED, record);
 
       matchingSubscriptions =
           subscriptionStore.findSubscriptions(message.getName(), message.getCorrelationKey());
 
       sideEffect.accept(this::correlateMessage);
 
-      messageStore.addMessage(message);
+      if (messageRecord.getTimeToLive() > 0L) {
+        message.setKey(key);
+        messageStore.addMessage(message);
+
+      } else {
+        // don't add the message to the store to avoid that it can be correlated afterwards
+        batchWriter.addFollowUpCommand(key, MessageIntent.DELETE, messageRecord);
```

Comment:
This is a bit inconsistent in my opinion. The message is never added to the store, so it is not required to delete it in a separate step. Also, having the extra asynchronous command suggests that the message can be correlated to subscriptions that are opened between `PUBLISH` and `DELETE`. I think it's cleaner if we either directly write the `DELETED` event or omit this record entirely.